File: plugin-sdk-ts/src/git.ts
/**
 * Git repository operations module.
 *
 * Provides read-only access to Git repository information such as
 * status, branches, commits, diffs, and contributors.
 * Limited write operations: stage files, commit.
 *
 * Requires: `git_read` and/or `git_write` permissions.
 */
import { callHostJson } from './host';
import type {
  GitRepoInfo,
  GitStatus,
  GitBranch,
  GitTag,
  GitCommit,
  GitCommitDetail,
  GitBlameEntry,
  GitRemote,
  GitStash,
  GitContributor,
  GitAheadBehind,
} from './types';

/** Check if git is available on the system. Requires: git_read */
export function isAvailable(): boolean {
  return callHostJson<boolean>('cognia_git_is_available', '');
}

/** Get git version string. Requires: git_read */
export function getVersion(): string {
  return callHostJson<string>('cognia_git_get_version', '');
}

/** Get repository information. Requires: git_read */
export function getRepoInfo(path: string): GitRepoInfo {
  return callHostJson<GitRepoInfo>(
    'cognia_git_get_repo_info',
    JSON.stringify({ path }),
  );
}

/** Get repository working tree status. Requires: git_read */
export function getStatus(path: string): GitStatus {
  return callHostJson<GitStatus>(
    'cognia_git_get_status',
    JSON.stringify({ path }),
  );
}

/** List branches. Requires: git_read */
export function getBranches(path: string): GitBranch[] {
  return callHostJson<GitBranch[]>(
    'cognia_git_get_branches',
    JSON.stringify({ path }),
  );
}

/** Get the current branch name. Requires: git_read */
export function getCurrentBranch(path: string): string | null {
  return callHostJson<string | null>(
    'cognia_git_get_current_branch',
    JSON.stringify({ path }),
  );
}

/** List tags. Requires: git_read */
export function getTags(path: string): GitTag[] {
  return callHostJson<GitTag[]>(
    'cognia_git_get_tags',
    JSON.stringify({ path }),
  );
}

/** Get commit history. Requires: git_read */
export function getLog(path: string, limit?: number, branch?: string): GitCommit[] {
  return callHostJson<GitCommit[]>(
    'cognia_git_get_log',
    JSON.stringify({ path, limit: limit ?? null, branch: branch ?? null }),
  );
}

/** Get detailed commit information. Requires: git_read */
export function getCommitDetail(path: string, hash: string): GitCommitDetail {
  return callHostJson<GitCommitDetail>(
    'cognia_git_get_commit_detail',
    JSON.stringify({ path, hash }),
  );
}

/** Get blame information for a file. Requires: git_read */
export function getBlame(path: string, file: string): GitBlameEntry[] {
  return callHostJson<GitBlameEntry[]>(
    'cognia_git_get_blame',
    JSON.stringify({ path, file }),
  );
}

/** Get working tree diff. Requires: git_read */
export function getDiff(path: string): string {
  return callHostJson<string>(
    'cognia_git_get_diff',
    JSON.stringify({ path }),
  );
}

/** Get diff between two revisions. Requires: git_read */
export function getDiffBetween(path: string, from: string, to: string): string {
  return callHostJson<string>(
    'cognia_git_get_diff_between',
    JSON.stringify({ path, from, to }),
  );
}

/** List remote repositories. Requires: git_read */
export function getRemotes(path: string): GitRemote[] {
  return callHostJson<GitRemote[]>(
    'cognia_git_get_remotes',
    JSON.stringify({ path }),
  );
}

/** List stashed changes. Requires: git_read */
export function getStashes(path: string): GitStash[] {
  return callHostJson<GitStash[]>(
    'cognia_git_get_stashes',
    JSON.stringify({ path }),
  );
}

/** Get repository contributors. Requires: git_read */
export function getContributors(path: string): GitContributor[] {
  return callHostJson<GitContributor[]>(
    'cognia_git_get_contributors',
    JSON.stringify({ path }),
  );
}

/** Search commits by query. Requires: git_read */
export function searchCommits(path: string, query: string): GitCommit[] {
  return callHostJson<GitCommit[]>(
    'cognia_git_search_commits',
    JSON.stringify({ path, query }),
  );
}

/** Get file commit history. Requires: git_read */
export function getFileHistory(path: string, file: string): GitCommit[] {
  return callHostJson<GitCommit[]>(
    'cognia_git_get_file_history',
    JSON.stringify({ path, file }),
  );
}

/** Get ahead/behind count relative to upstream. Requires: git_read */
export function getAheadBehind(path: string, branch?: string): GitAheadBehind {
  return callHostJson<GitAheadBehind>(
    'cognia_git_get_ahead_behind',
    JSON.stringify({ path, branch: branch ?? null }),
  );
}

/** Stage specific files. Requires: git_write */
export function stageFiles(path: string, files: string[]): void {
  callHostJson<{ ok: boolean }>(
    'cognia_git_stage_files',
    JSON.stringify({ path, files }),
  );
}

/** Create a commit. Requires: git_write */
export function commit(path: string, message: string): string {
  return callHostJson<string>(
    'cognia_git_commit',
    JSON.stringify({ path, message }),
  );
}

File: plugin-sdk/src/git.rs
//! Git repository operations module.
//!
//! Provides read-only access to Git repository information and limited write operations.

use crate::host;
use crate::types::*;
use extism_pdk::*;

/// Check if git is available. Requires: git_read
pub fn is_available() -> Result<bool, Error> {
    let result = unsafe { host::cognia_git_is_available(String::new())? };
    Ok(serde_json::from_str(&result)?)
}

/// Get git version string. Requires: git_read
pub fn get_version() -> Result<String, Error> {
    let result = unsafe { host::cognia_git_get_version(String::new())? };
    Ok(serde_json::from_str(&result)?)
}

/// Get repository information. Requires: git_read
pub fn get_repo_info(path: &str) -> Result<GitRepoInfo, Error> {
    let input = serde_json::json!({ "path": path }).to_string();
    let result = unsafe { host::cognia_git_get_repo_info(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Get repository working tree status. Requires: git_read
pub fn get_status(path: &str) -> Result<GitStatus, Error> {
    let input = serde_json::json!({ "path": path }).to_string();
    let result = unsafe { host::cognia_git_get_status(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// List branches. Requires: git_read
pub fn get_branches(path: &str) -> Result<Vec<GitBranch>, Error> {
    let input = serde_json::json!({ "path": path }).to_string();
    let result = unsafe { host::cognia_git_get_branches(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Get the current branch name. Requires: git_read
pub fn get_current_branch(path: &str) -> Result<Option<String>, Error> {
    let input = serde_json::json!({ "path": path }).to_string();
    let result = unsafe { host::cognia_git_get_current_branch(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// List tags. Requires: git_read
pub fn get_tags(path: &str) -> Result<Vec<GitTag>, Error> {
    let input = serde_json::json!({ "path": path }).to_string();
    let result = unsafe { host::cognia_git_get_tags(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Get commit history. Requires: git_read
pub fn get_log(path: &str, limit: Option<u32>, branch: Option<&str>) -> Result<Vec<GitCommit>, Error> {
    let input = serde_json::json!({ "path": path, "limit": limit, "branch": branch }).to_string();
    let result = unsafe { host::cognia_git_get_log(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Get detailed commit information. Requires: git_read
pub fn get_commit_detail(path: &str, hash: &str) -> Result<GitCommitDetail, Error> {
    let input = serde_json::json!({ "path": path, "hash": hash }).to_string();
    let result = unsafe { host::cognia_git_get_commit_detail(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Get blame information for a file. Requires: git_read
pub fn get_blame(path: &str, file: &str) -> Result<Vec<GitBlameEntry>, Error> {
    let input = serde_json::json!({ "path": path, "file": file }).to_string();
    let result = unsafe { host::cognia_git_get_blame(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Get working tree diff. Requires: git_read
pub fn get_diff(path: &str) -> Result<String, Error> {
    let input = serde_json::json!({ "path": path }).to_string();
    let result = unsafe { host::cognia_git_get_diff(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Get diff between two revisions. Requires: git_read
pub fn get_diff_between(path: &str, from: &str, to: &str) -> Result<String, Error> {
    let input = serde_json::json!({ "path": path, "from": from, "to": to }).to_string();
    let result = unsafe { host::cognia_git_get_diff_between(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// List remote repositories. Requires: git_read
pub fn get_remotes(path: &str) -> Result<Vec<GitRemote>, Error> {
    let input = serde_json::json!({ "path": path }).to_string();
    let result = unsafe { host::cognia_git_get_remotes(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// List stashed changes. Requires: git_read
pub fn get_stashes(path: &str) -> Result<Vec<GitStash>, Error> {
    let input = serde_json::json!({ "path": path }).to_string();
    let result = unsafe { host::cognia_git_get_stashes(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Get repository contributors. Requires: git_read
pub fn get_contributors(path: &str) -> Result<Vec<GitContributor>, Error> {
    let input = serde_json::json!({ "path": path }).to_string();
    let result = unsafe { host::cognia_git_get_contributors(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Search commits by query. Requires: git_read
pub fn search_commits(path: &str, query: &str) -> Result<Vec<GitCommit>, Error> {
    let input = serde_json::json!({ "path": path, "query": query }).to_string();
    let result = unsafe { host::cognia_git_search_commits(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Get file commit history. Requires: git_read
pub fn get_file_history(path: &str, file: &str) -> Result<Vec<GitCommit>, Error> {
    let input = serde_json::json!({ "path": path, "file": file }).to_string();
    let result = unsafe { host::cognia_git_get_file_history(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Get ahead/behind count relative to upstream. Requires: git_read
pub fn get_ahead_behind(path: &str, branch: Option<&str>) -> Result<GitAheadBehind, Error> {
    let input = serde_json::json!({ "path": path, "branch": branch }).to_string();
    let result = unsafe { host::cognia_git_get_ahead_behind(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Stage specific files. Requires: git_write
pub fn stage_files(path: &str, files: &[&str]) -> Result<(), Error> {
    let input = serde_json::json!({ "path": path, "files": files }).to_string();
    unsafe { host::cognia_git_stage_files(input)?; }
    Ok(())
}

/// Create a commit. Returns commit hash. Requires: git_write
pub fn commit(path: &str, message: &str) -> Result<String, Error> {
    let input = serde_json::json!({ "path": path, "message": message }).to_string();
    let result = unsafe { host::cognia_git_commit(input)? };
    Ok(serde_json::from_str(&result)?)
}
